File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.js
/**
 * @param {string} s
 * @return {number}
 */
var distinctSubseqII = function(s) {
  const end = new Array(26).fill(0);
  const mod = 1e9 + 7;
  let ans = 0;
  
  for (let ch of s) {
    const i = ch.charCodeAt() - 'a'.charCodeAt();
    const others = (ans - end[i] + mod) % mod;
    end[i] = ans + 1;
    ans = (end[i] + others) % mod;
  }
  
  return ans;
};

File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        end = [0] * 26
        MOD, ans = 10**9 + 7, 0
        for c in s:
            i = ord(c) - ord('a')
            ans, end[i] = ans * 2 + 1 - end[i], ans + 1
        return ans % MOD
      